certinfo.go

```diff
@@ -749,20 +844,218 @@ func CertificateRequestText(csr *x509.CertificateRequest) (string, error) {
+					buf.WriteString(fmt.Sprintf("%16sPermitted:\n%18s%s", "", "", permittedDNSDomains[0]))
+					for i := 1; i < len(permittedDNSDomains); i++ {
+						buf.WriteString(fmt.Sprintf(", %s", permittedDNSDomains[i]))
+					}
+					buf.WriteString("\n")
+				}
+			case ext.Id.Equal(oidExtExtendedKeyUsage):
+				// extKeyUsage: RFC 5280, 4.2.1.12
+				// id-ce-extKeyUsage OBJECT IDENTIFIER ::= { id-ce 37 }
+				//
+				// ExtKeyUsageSyntax ::= SEQUENCE SIZE (1..MAX) OF KeyPurposeId
+				//
+				// KeyPurposeId ::= OBJECT IDENTIFIER
+				var keyUsage []asn1.ObjectIdentifier
+				_, err = asn1.Unmarshal(ext.Value, &keyUsage)
+				if err != nil {
+					continue
+				}
+				var eku []x509.ExtKeyUsage
+				for _, u := range keyUsage {
+					if extKeyUsage, ok := extKeyUsageFromOID(u); ok {
+						eku = append(eku, extKeyUsage)
+					}
+				}
+				buf.WriteString(fmt.Sprintf("%12sX509v3 Extended Key Usage:", ""))
+				if ext.Critical {
+					buf.WriteString(" critical\n")
+				} else {
+					buf.WriteString("\n")
+				}
+				var list []string
+				for _, val := range eku {
+					switch val {
+					case x509.ExtKeyUsageAny:
+						list = append(list, "Any Usage")
+					case x509.ExtKeyUsageServerAuth:
+						list = append(list, "TLS Web Server Authentication")
+					case x509.ExtKeyUsageClientAuth:
+						list = append(list, "TLS Web Client Authentication")
+					case x509.ExtKeyUsageCodeSigning:
+						list = append(list, "Code Signing")
+					case x509.ExtKeyUsageEmailProtection:
+						list = append(list, "E-mail Protection")
+					case x509.ExtKeyUsageIPSECEndSystem:
+						list = append(list, "IPSec End System")
+					case x509.ExtKeyUsageIPSECTunnel:
+						list = append(list, "IPSec Tunnel")
+					case x509.ExtKeyUsageIPSECUser:
+						list = append(list, "IPSec User")
+					case x509.ExtKeyUsageTimeStamping:
+						list = append(list, "Time Stamping")
+					case x509.ExtKeyUsageOCSPSigning:
+						list = append(list, "OCSP Signing")
+					default:
+						list = append(list, "UNKNOWN")
+					}
+				}
+				if len(list) > 0 {
+					buf.WriteString(fmt.Sprintf("%16s%s", "", list[0]))
+					for i := 1; i < len(list); i++ {
+						buf.WriteString(fmt.Sprintf(", %s", list[i]))
+					}
+					buf.WriteString("\n")
+				}
+			default:
+				unknownExts = append(unknownExts, ext)
 			}
 			if err != nil {
 				return "", err
 			}
 		}
+		if len(unknownExts) > 0 {
+			buf.WriteString(fmt.Sprintf("%8sAttributes:\n", ""))
+			for _, ext := range unknownExts {
+				buf.WriteString(fmt.Sprintf("%12s%s:", "", ext.Id.String()))
+				if ext.Critical {
+					buf.WriteString(" critical\n")
+				} else {
+					buf.WriteString("\n")
+				}
+				value := bytes.Runes(ext.Value)
+				sanitized := make([]rune, len(value))
+				for i, r := range value {
+					if strconv.IsPrint(r) && r != '�' {
+						sanitized[i] = r
+					} else {
+						sanitized[i] = '.'
+					}
+				}
+				buf.WriteString(fmt.Sprintf("%16s%s\n", "", string(sanitized)))
```

Comment:
If it's not printable we can print the hexadecimal version of the bytes.

Reply:
Are you saying if `buf.WriteString` returns an error?

Reply:
No, few lines above you're checking the characters in the value of the extension, in some cases they will be strings, but most of the time they will be bytes that we cannot interpret. Writing them in hexadecimal encoding will allow the user to decode them if he knows how.

Reply:
So are you saying that if `value` has any unprintable characters, then print the entirety of `value` in hexadecimal. Or only printing those characters in hexadecimal that are not printable.

e.g. 

```
Attributes:
            1.2.3.4.5:
                foo
            1.1.13.1.2.4.15.17.1.3.1.2.4.1: critical
                0...max..furman
```

So that would be mean printing the entirety of the line w/ `max..furman` in hexadecimal, right?

Reply:
We can print both, if it has unprintable characters.